src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -68,187 +66,194 @@ ShardManager::Result< ShardInfo > HSHomeObject::_create_shard(pg_id_t pg_owner,
+    if (hs_ctx != nullptr) {
+        auto ctx = boost::static_pointer_cast< repl_result_ctx< ShardManager::Result< ShardInfo > > >(hs_ctx);
+        do_shard_message_commit(lsn, *r_cast< ReplicationMessageHeader* >(header.bytes), blkids, ctx->hdr_buf_, hs_ctx);
         return;
     }
 
-    std::scoped_lock lock_guard(_flying_shard_lock);
-    auto iter = _flying_shards.find(lsn);
-    if (iter != _flying_shards.end()) { _flying_shards.erase(iter); }
+    // hs_ctx will be nullptr when HS is restarting and replay all commited log entries from the last checkpoint;
+    // but do we really need to handle this for create_shard or seal_shard?
+    // because this raft log had been commit before HO restarts and commit result is already saved in HS metablks
+    // when HS restarts, all PG/Shard infos will be recovered from HS metablks and if we commit again, it will cause
+    // duplication;
+#if 0     
+    sisl::sg_list value;
+    value.size = blkids.blk_count() * repl_dev->get_blk_size();
+    auto value_buf = iomanager.iobuf_alloc(512, value.size);
+    value.iovs.push_back(iovec{.iov_base = value_buf, .iov_len = value.size});
+    // header will be released when this function returns, but we still need the header when async_read() finished.
+    ReplicationMessageHeader msg_header = *r_cast< const ReplicationMessageHeader* >(header.bytes);
+    repl_dev->async_read(blkids, value, value.size).thenValue([this, lsn, msg_header, blkids, value](auto&& err) {
+        if (err) {
+            LOGW(homeobject, "failed to read data from homestore pba, lsn:{}", lsn);
+        } else {
+            sisl::blob value_blob(r_cast< uint8_t* >(value.iovs[0].iov_base), value.size);
+            do_shard_message_commit(lsn, msg_header, blkids, value_blob, nullptr);
+        }
+        iomanager.iobuf_free(r_cast< uint8_t* >(value.iovs[0].iov_base));
+    });
+#endif
 }
 
-void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& header, sisl::blob const& key,
+void HSHomeObject::do_shard_message_commit(int64_t lsn, ReplicationMessageHeader& header,
+                                           homestore::MultiBlkId const& blkids, sisl::blob value,
+
                                            cintrusive< homestore::repl_req_ctx >& hs_ctx) {
-    // ctx will be nullptr when:
-    // 1. on the follower side
-    // 2. on the leader side but homeobject restarts and replay all commited log entries from the last commit id;
     repl_result_ctx< ShardManager::Result< ShardInfo > >* ctx{nullptr};
     if (hs_ctx != nullptr) {
         ctx = boost::static_pointer_cast< repl_result_ctx< ShardManager::Result< ShardInfo > > >(hs_ctx).get();
     }
 
-    std::string shard_msg;
-    if (!precheck_and_decode_shard_msg(lsn, header, key, &shard_msg)) {
+    if (header.corrupted()) {
+        LOGW("replication message header is corrupted with crc error, lsn:{}", lsn);
+        if (ctx) { ctx->promise_.setValue(folly::makeUnexpected(ShardError::INVALID_ARG)); }
```

Comment:
Its not invalid arg from user. A new error HEADER or MSG CORRUPT ?

Reply:
`ShardError::CRC_MISMATCH`?

Reply:
Agreed.